src/main/scala/chisel3/internal/firrtl/Converter.scala

```diff
@@ -0,0 +1,238 @@
+
+  // alt indicates to a WhenEnd whether we're closing an alt or just a regular when
+  // TODO we should probably have a different structure in the IR to close elses
+  private case class WhenFrame(when: fir.Conditionally, outer: Queue[fir.Statement], alt: Boolean)
+
+  // Whens markers are flat so scope must be inferred
+  // TODO refactor
+  def convert(cmds: Seq[Command], ctx: Component): fir.Statement = {
+    @tailrec
+    def rec(acc: Queue[fir.Statement],
+            scope: List[WhenFrame])
+           (cmds: Seq[Command]): Seq[fir.Statement] = {
+      if (cmds.isEmpty) {
+        assert(scope.isEmpty)
+        acc
+      } else cmds.head match {
+        case e: DefPrim[_] =>
+          val consts = e.args.collect { case ILit(i) => i }
+          val args = e.args.flatMap {
+            case _: ILit => None
+            case other => Some(convert(other, ctx))
+          }
+          val expr = e.op.name match {
+            case "mux" =>
+              assert(args.size == 3, s"Mux with unexpected args: $args")
+              fir.Mux(args(0), args(1), args(2), fir.UnknownType)
+            case _ =>
+              fir.DoPrim(convert(e.op), args, consts, fir.UnknownType)
+          }
+          val node = fir.DefNode(convert(e.sourceInfo), e.name, expr)
+          rec(acc :+ node, scope)(cmds.tail)
+        case e @ DefWire(info, id) =>
+          val wire = fir.DefWire(convert(info), e.name, extractType(id))
+          rec(acc :+ wire, scope)(cmds.tail)
+        case e @ DefReg(info, id, clock) =>
+          val reg = fir.DefRegister(convert(info), e.name, extractType(id), convert(clock, ctx),
+                                    firrtl.Utils.zero, convert(id.getRef, ctx))
+          rec(acc :+ reg, scope)(cmds.tail)
+        case e @ DefRegInit(info, id, clock, reset, init) =>
+          val reg = fir.DefRegister(convert(info), e.name, extractType(id), convert(clock, ctx),
+                                    convert(reset, ctx), convert(init, ctx))
+          rec(acc :+ reg, scope)(cmds.tail)
+        case e @ DefMemory(info, id, t, size) =>
+          val mem = firrtl.CDefMemory(convert(info), e.name, extractType(t), size, false)
+          rec(acc :+ mem, scope)(cmds.tail)
+        case e @ DefSeqMemory(info, id, t, size) =>
+          val mem = firrtl.CDefMemory(convert(info), e.name, extractType(t), size, true)
+          rec(acc :+ mem, scope)(cmds.tail)
+        case e: DefMemPort[_] =>
+          val port = firrtl.CDefMPort(convert(e.sourceInfo), e.name, fir.UnknownType,
+            e.source.fullName(ctx), Seq(convert(e.index, ctx), convert(e.clock, ctx)), convert(e.dir))
+          rec(acc :+ port, scope)(cmds.tail)
+        case Connect(info, loc, exp) =>
+          val con = fir.Connect(convert(info), convert(loc, ctx), convert(exp, ctx))
+          rec(acc :+ con, scope)(cmds.tail)
+        case BulkConnect(info, loc, exp) =>
+          val con = fir.PartialConnect(convert(info), convert(loc, ctx), convert(exp, ctx))
```

Comment:
Are FIRRTL PartialConnect semantics equivalent to Chisel BulkConnect semantics? Or should this exception out because this should never be seen?

Reply:
FIRRTL PartialConnect semantics are closer to Chisel2 (and the current `Chisel._` `<>`) semantics actually. It's still used by the compatibility wrapper.